### PR TITLE
Implement the dusk contract circuits

### LINKED ISSUE #59: Generation of the variants for all the 5 unique circuits of Dusk Token
We need to have a `PreProcessedCircuit` for each variant of each circuit of the Dusk Token.

Once dusk-network/rusk#57 gets closed, we will have a basis with all of the gadgets tested for each of the circuits.

The idea then will be: Making usage of a Macro or a similar technique that allows us to not re-write code. (I envision this as a macro that iterates over the gadgets X amount of times depending on the amount of Notes that the Tx contains), implement a method that generates the 8 variants of each unique circuit required by the Dusk Token in respect of the amount of Notes.

Appart of implementing that method, we should also add a test that checks all of the variants correctness.

Suggestions:

- In the tests, just generate once the `PublicParameters` and ck & vk and make them big enough to support the 4INP 4OUT case. On that way, you can iterate over Transactions with different amounts of inputs with a different composer each time but keeping the same Trusted Setup so tests don't take a lot of time.

### PR DESCRIPTION
**Pull request by Bounce23**
The implementation of the dusk contract circuits will be done in this PR. These circuits, and their methodic format are documented in the [gitbook](https://app.gitbook.com/@dusk-network/s/specs/specifications/smart-contracts/genenesis-contracts/dusk-contract).

This closes #59. This issue lives inside epic #57.

### CODE REVIEW

**Comment by Bounce23** [comment]
Due to a need to change in the specs for two of the circuits, 4/6 have been completed. 
Those needing to be refactored when the specs are update, is the `Execute.rs` file and the `send_to_contract_obfuscated`. 
The change for those files will be the signature scheme and the encryption scheme, respectively. 

@CPerezz Would you rather I got this PR done with the circuit builder etc for the 4 working circuits and merged. Then opened issues so I can work on the Regular token circuits and open an issue to update the other circuits when the specs are ready. _(This is similar to how it was done last time)_. **Or** put this issue as blocked? @ZER0, perhaps you have a weigh in on how is best to structure the workflow.

File: contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs
Comment (julesdesmit) [inline contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs:30]: Make sure the comments for `pi_constructor` and `size` are doc comments.

**Review by julesdesmit — CHANGES_REQUESTED** [review]
Just one small nitpick - from my end, the logic LGTM, but I'll defer to @CPerezz to be the judge of that.

File: contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs
Comment (Bounce23) [reply contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs:30]: Done.

File: contracts/transfer/circuits/Cargo.toml
Comment (CPerezz) [inline contracts/transfer/circuits/Cargo.toml:20]: Why not moving this to [dependencies] directly?

File: contracts/transfer/circuits/src/dusk_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract.rs:12]: Instead of importing the modules publicly. Import them without the `pub` and do a `pub use` for the specific Circuit implementations. Will look cleaner.

File: contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs:86]: Does it need to be that big? I don't think so.

File: contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs:147]: Same as avobe.

File: contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs:165]: Same as avobe.

File: contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs:208]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs:247]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/send_to_contract_transparent.rs:288]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs:37]: Is there any structure that holds more than one of these fields inside? If that's the case, you should add the structure here instead of its fields.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs:147]: Too big. The circuit is for sure smaller, so there's no need to trim that big.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs:208]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs:226]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs:345]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated.rs:288]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs:149]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs:210]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs:228]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs:351]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs:492]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs:549]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs:603]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs:653]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs:707]: Same as above.

File: contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/withdraw_from_obfuscated_to_contract.rs:730]: Remove this empty lines please

**Review by CPerezz — CHANGES_REQUESTED** [review]
Apply `rustfmt`

File: contracts/transfer/circuits/src/dusk_contract/execute.rs
Comment (julesdesmit) [inline contracts/transfer/circuits/src/dusk_contract/execute.rs:572]: You seem to be repeating this logic quite often in multiple tests multiple times, so I think this is a good candidate as well for abstracting away into a function. It also seems like most of the stuff you declare you won't need later on, so it could possibly look like this:

```rust
fn circuit_note(ssk: &SecretSpendKey, value: u64, pos: u64, blinder: &JubJubScalar) -> Note {
    let r1 = JubJubScalar::from(150 as u64);
    let nonce1 = JubJubScalar::from(350 as u64);
    let psk1 = PublicSpendKey::from(ssk1);
    let note = Note::deterministic(
        NoteType::Transparent,
        &r1,
        nonce1,
        &psk1,
        value1,
        input_note_blinder_one,
    );
    note.set_pos(pos);
    note
}
```

This removes the declaration of variables that are not needed later on, outside of the test. Additionally, you seem to be declaring variables which are kind of unnecessary:

```rust
let pos1 = note1.set_pos(0);
// Define the hash of the note, which is required for the preimage check
let note_hash1 = note1.hash();
let pos_a = note1.pos();
// Derive the one time secret key, sk_r, for the note
let sk_r1 = ssk1.sk_r(note1.stealth_address());
// Derive the one time public key, pk_r, for the note
let pk_r1 = note1.stealth_address().pk_r();
```

When constructing the circuit later on, you can directly reference these things from the note you've made with the `circuit_note` function, which saves some lines of code and helps make this easier to read.

```rust
let mut circuit = build_execute_circuit(
    vec![note1.gen_nullifier(&ssk1), note2.gen_nullifier(&ssk2)],
    vec![note1.hash(), note2.hash()],
...
);
```

Finally, when you re-use a value, for instance, with the `input_note_value_one` variable, it is a good rule of thumb to only declare it once and then re-use that variable, to avoid potential mistakes, and aid readability. For instance:

```rust
let value1 = 600u64;
...
let input_note_value_one = JubJubScalar::from(value1);
```

Even more concise, would be to do this:

```rust
let input_commitment_one = compute_value_commitment(
    JubJubScalar::from(value1),
    input_note_blinder_one,
);
```

All things combined, you can cut down on a LOT of code this way, which will bring lots of benefits for us in the future when we have to maintain this code.

File: contracts/transfer/circuits/src/dusk_contract/execute.rs
Comment (julesdesmit) [inline contracts/transfer/circuits/src/dusk_contract/execute.rs:705]: This seems to be another candidate for reducing code repetition. You can do something like:

```rust
fn add_circuit_public_inputs(circuit: &ExecuteCircuit, pi: Vec<PublicInput>) {
    circuit.
        input_poseidon_branches
...
}
```

**Review by julesdesmit — CHANGES_REQUESTED** [review]
Just some readability nits on the execute circuit, the rest looks fine to me.

**Review by CPerezz — CHANGES_REQUESTED** [review]
In respect to the Execute circuit, I would like to see comments that specify where you set the incorrect parameters and how this affects the circuit. Like you did in `WithdrawFromObfuscatedToContractCircuitOne`

File: contracts/transfer/circuits/src/dusk_contract/execute.rs
Comment (Bounce23) [reply contracts/transfer/circuits/src/dusk_contract/execute.rs:617]: It happens to be those values but I picked them due to ease. They must satisfy `input_note_value - output_note_value - crossover_value - fee = 0` and its 800 - 400 - 200 - 200 but it could also be 900 - 100 - 300 - 500. So they wouldn't be reusable.

File: contracts/transfer/circuits/src/dusk_contract/execute.rs
Comment (julesdesmit) [inline contracts/transfer/circuits/src/dusk_contract/execute.rs:623]: Please abstract the note creation away into a function like previously mentioned - make sure as well, that you delete the variables here which are now unused, due to direct referencing of the note.

**Review by julesdesmit — CHANGES_REQUESTED** [review]

**Review by julesdesmit — APPROVED** [review]
LGTM, let's see what @CPerezz has to say before merging.

**Review by CPerezz — APPROVED** [review]
LGTM

File: contracts/transfer/circuits/src/dusk_contract/execute.rs
Comment (CPerezz) [inline contracts/transfer/circuits/src/dusk_contract/execute.rs:307]: Handle this access properly so that we don't panic here.

**Review by CPerezz — APPROVED** [review]
LGTM. Just two things before merging:
- Address the panic handling mentioned.
- Rebase the commits so that we don't pollute the history that much.

**Pull request merged**
